packages/prover/src/proof_provider/proof_provider.ts

```diff
@@ -0,0 +1,182 @@
+import {assertLightClient} from "../utils/assertion.js";
+import {
+  getExecutionPayloads,
+  getGenesisData,
+  getSyncCheckpoint,
+  getUnFinalizedRangeForPayloads,
+} from "../utils/consensus.js";
+import {PayloadStore} from "./payload_store.js";
+
+type RootProviderOptions = Omit<RootProviderInitOptions, "transport"> & {
+  transport: LightClientRestTransport;
+  api: Api;
+  config: ChainForkConfig;
+};
+
+export class ProofProvider {
+  private store: PayloadStore;
+
+  private readyPromise?: Promise<void>;
+  lightClient?: Lightclient;
+
+  constructor(private opts: RootProviderOptions) {
+    this.store = new PayloadStore({api: opts.api});
+  }
+
+  async waitToBeReady(): Promise<void> {
+    return this.readyPromise;
+  }
+
+  static init(opts: RootProviderInitOptions): ProofProvider {
+    if (opts.transport === LCTransport.P2P) {
+      throw new Error("P2P mode not supported yet");
+    }
+
+    const config = createChainForkConfig(networksChainConfig[opts.network]);
+    const api = getClient({urls: opts.urls}, {config});
+    const transport = new LightClientRestTransport(api);
+
+    const provider = new ProofProvider({
+      ...opts,
+      config,
+      api,
+      transport,
+    });
+
+    provider.readyPromise = provider.sync(opts.wsCheckpoint);
+
+    return provider;
+  }
+
+  private async sync(wsCheckpoint?: string): Promise<void> {
+    if (this.lightClient !== undefined) {
+      throw Error("Light client already initialized and syncing.");
+    }
+
+    const {api, config, transport} = this.opts;
+    const checkpointRoot = await getSyncCheckpoint(api, wsCheckpoint);
+    const genesisData = await getGenesisData(api);
+
+    this.lightClient = await Lightclient.initializeFromCheckpointRoot({
+      checkpointRoot,
+      config,
+      transport,
+      genesisData,
+    });
+
+    assertLightClient(this.lightClient);
+    // Wait for the lightclient to start
+    await new Promise<void>((resolve) => {
+      const lightClientStarted = (status: RunStatusCode): void => {
+        if (status === RunStatusCode.started) {
+          this.lightClient?.emitter.off(LightclientEvent.statusChange, lightClientStarted);
+          resolve();
+        }
+      };
+      this.lightClient?.emitter.on(LightclientEvent.statusChange, lightClientStarted);
+      this.lightClient?.start();
+    });
+    this.registerEvents();
+
+    // Load the payloads from the CL
+    const {start, end} = await getUnFinalizedRangeForPayloads(this.lightClient);
+    const payloads = await getExecutionPayloads(this.opts.api, start, end);
+    for (const payload of Object.values(payloads)) {
+      this.store.set(payload, false);
+    }
+
+    // Load the finalized payload from the CL
+    const finalizedSlot = this.lightClient.getFinalized().beacon.slot;
+    const finalizedPayload = await getExecutionPayloads(this.opts.api, finalizedSlot, finalizedSlot);
+    this.store.set(finalizedPayload[finalizedSlot], true);
```

Comment:
you can use a better pattern of Object.values(),map(payload=>this.store.set(payload)) here and everywhere else

Reply:
`Object.values` is used where it's intended use. On this line we need to set a single payload so didn't used it.